### ISSUE TITLE
git lfs dedup reports system not supported even though it should work

### ISSUE THREAD

**Issue by jochenhz** [git-lfs/git-lfs#3970]
Hey there,

I was looking at the git lfs dedup command which sounds really useful, thanks for working on that. I am running macOS 10.14.6 on an APFS volume (encrypted). 
git lfs dedup --test reports : "This system does not support deduplication. unsupported platform"

Running cp -c in terminal reports no error so I am assuming clonefile is working correctly (I could not figure out how to determine clonefile actually working on macOS)

I am running: 
git-lfs/2.9.2 (GitHub; darwin amd64; go 1.12.7; git 0274d856)

Any suggestions? Am I missing something?
Thanks a lot!

**Comment by bk2204** [comment]
Hey,

How did you get the binary that you're using?  I believe this operation requires that the Darwin binary be built on Darwin itself, and if you're using a binary that was build on another OS (like our distributed binary), it may not work properly.

This is a limitation of CGo and cross-compiling.  If you're using our binaries, we'll fix that when we start signing binaries for macOS.

**Comment by J0Nes90** [comment]
Hey, thanks for your answer. 
The first binary I got from the Git LFS webpage: https://github.com/git-lfs/git-lfs/releases/download/v2.9.1/git-lfs-darwin-amd64-v2.9.1.tar.gz

Later, I upgraded to 2.9.2 using this link: https://github.com/git-lfs/git-lfs/releases/download/v2.9.2/git-lfs-darwin-amd64-v2.9.2.tar.gz

To be clear, what binary should I be using? Isn't Darwin the correct one for macOS? Should I compile it myself?

**Comment by bk2204** [comment]
The problem isn't that you downloaded the wrong binary; you did download the correct one.  The limitation is how we build our official binaries at the moment.

For this feature to work on Darwin and Linux, the binary has to be compiled on the same type of system as it's supposed to be run on, because it uses cgo, which requires support from the C toolchain and headers on that system.  If the binary is compiled on a different system, it doesn't use cgo, and we compile in a stub implementation that doesn't do anything.  That implementation provides the message you're seeing.

Currently, we build Windows binaries on Windows since we need to sign them, and we build all other binaries on Linux (because it's fast and consistent).  We'll start signing our macOS binaries, and when we do, we'll build them on macOS, at which point we'll have cgo available, and the full implementation will be compiled in automatically.

In the mean time, you can either compile the binary yourself, or you can use the Homebrew version, which should work automatically (having been compiled on macOS).

**Comment by J0Nes90** [comment]
Alright, I understand. Thanks a lot for the clarification!
I can confirm that using the version from homebrew is fixing the issue for me.

I see that using the dedup command reports lots of de-duplicated bytes (even though the OS still fails to report them as clones)

I'd like to use the dedup to actually reduce the amount of used space by the cache / worktree (c.f. https://github.com/git-lfs/git-lfs/issues/3635). I was wondering what drawbacks I'd face by just deleting the .git/lfs/objects folder entirely. So far the only notable drawback is that I cannot revert a change on HEAD without an active connection to the server. Is there anything I am missing?

Again, thank you very much!

**Comment by bk2204** [comment]
It's not recommended to delete the folder entirely, because any objects that you've created but haven't been pushed will be destroyed.  I'd recommend running `git lfs prune` instead to reduce the amount of space, and possibly tuning the parameters to fit your workload better.

**Comment by J0Nes90** [comment]
That makes a lot of sense, thanks for your recommendations. I can reproduce push not working when deleting unpushed files in git/lfs/objects.
With that being true one can assume deleting objects in .git/lfs/objects to be safe only and only if they are pushed to a remote. Wouldn't it be nice to add an option to `git lfs prune` that prunes **everything** that is pushed (e.g. with `--verify-remote` being enabled by default)? Even by tuning parameters and configurations from the prune command does at least double the required disk space of my repository.

**Comment by bk2204** [comment]
I don't think we want to prune things that are used by the current references, because it makes it hard to do a `git reset --hard` or otherwise reset files without downloading objects.

I'm not sure why the OS fails to report your files as clones.  Perhaps @kazuki-ma, as the original author, might know more.  I think fixing that issue would probably get your space savings without the need for deleting objects required by the current reference.

**Comment by J0Nes90** [comment]
Right, by doing that you'd lose the ability to `git reset ` without downloading files again. In my situation that is totally acceptable, however (using a local server). Of course, one needs to know what he is doing in that case. 
Our repositories can become pretty massive, unfortunately. A lot of the files in the repo are read only for the current project, so having them sit in .git/lfs/objects is wasted space.  I am sure that there is a good use case not only for my purpose.

Of course, the dedup functionality would be a workaround (at least on macOS, not sure if we convince our IT to move our filesystem to ReFS on windows). Such a prune command would be an expert option, but a very welcome one.

Thanks a lot for your support. Maybe @kazuki-ma has more insights how to check if the dedup on macOS did actually work.

**Comment by kazuki-ma** [comment]
thank you for notify me. I'll looking it.

**Comment by bk2204** [comment]
I think we'll need to see a testcase to figure out why this isn't working for de-duplication on macOS.  As such, I'm going to mark it to see if we can get some reproduction steps.

**Comment by chrisd8088** [comment]
With both our v3.6.1 release build of the Git LFS client and the corresponding Homebrew build, `git lfs dedup` appears to work as expected on recent macOS versions:
```
$ git lfs dedup --test
OK: This platform and repository support file de-duplication.

$ git lfs dedup
...
Finished successfully.
  De-duplicated  size: 108 bytes
                count: 9
```

I'm therefore going to close this issue, but if it's still a problem for anyone, please re-open it and provide a full reproduction case and the details of your OS and Git LFS versions.
